lib/pure/algorithm.nim

```diff
@@ -369,3 +369,113 @@ when isMainModule:
   for i in 0 .. high(arr1):
     assert arr1.reversed(0, i) == arr1.reversed()[high(arr1) - i .. high(arr1)]
     assert arr1.reversed(i, high(arr1)) == arr1.reversed()[0 .. high(arr1) - i]
+
+
+proc rotate*[T](arg: var openarray[T]; first, middle, last: int): int {.discardable.} =
+  ## Performs a left rotation on a range of elements.
+  ## Specifically, ``rotate`` swaps the elements in the range [first, last) in such a way
+  ## that the element at index ``middle`` becomes the first element of the sub range and
+  ## ``middle - 1`` becomes the last element of the sub range. The time complexity is
+  ## linear to ``last - first``. returns ``first + (last - middle)``. ``T`` must support swap operation
+  ##
+  ## ``first``
+  ##   the index of the first element that should be rotated.
+  ##
+  ## ``middle``
+  ##   the index of the element that should appear at the beginning of the rotated range.
+  ##
+  ## ``last``
+  ##   the index of the first element that should not be rotated anymore (exclusive upper bound).
```

Comment:
From my c++ past, I see indices more as a separator that sits between elements like a vertical bar write cursor in most text editors. Inclusive upper bounds are more like the vim cursor in normal mode, where I have to press a different key to enter the insert mode at the end of a line unlike at the beginning of a line. But I agree that it should be consistent with the rest of the standard library.


    a b c d c f|g h i j k l m n|o p q r s t u|v w x y z
               |               |             |
             begin           middle         end

As I am writing this I finally understood waht you meant. I have used enums so far only in ``case of`` branches, so what I am saying here is just intuition, and could be total bogus. 

* `len(X)` is the amount of elements of `X`
* -> `high(X)` is the last valid index into `X`, therefore just an alias for `len(X)-1`. 
* -> `high(myenum)+1 === len(myenum)-1+1 === len(myenum)`
* -> so `high(myenum)+1` is just an integer with the amount of values from that enum

But this only makes sense when the enum is dense.
.